---
title: "Final Report"
author: "Alexander Karr"
date: "2024-04-23"
output: 
  pdf_document:
    toc: true
    toc_depth: 2  # Adjust depth as needed
---

# Introduction:
Despite the OSKM reprogramming factors having been initially reported as inducing genome-wide transcriptional responses, actually, less than 1% of cells undergo *authentic* pluripotency reprogramming (become iPSCs). In reality, their pluripotency induction is extremely inefficient, slow, and stochastic. To develop a sense of what kinds of transcriptomic changes the Yamanaka factors struggle to induce, and consistently succeed at, the authors of the paper that most inspired me - called Human Transcription factor responsive to initial reprogrmaming predominantly undergo legitimate reprogramming during fibroblast conversion to iPSCs - developed the concept of the “reprogramome” – or the full complement of genes requiring reprogramming for a differentiated cell to become an iPSC. Reprogamming letigimacy, then, describes scenarios where genes fully reach iPSC expression patterns.
One of the most essential steps towards pluripotency induction is erasing the differentiated cell’s transcription factor system, and inducing the pluripotent TF system. These transcription factors can be thought of in four different categories in terms of their response to OSKM induction - in their progression from a differentiated state to that of an iPSC. The authors of the paper that most inspired me - *Human transcription factors responsive to initial reprogramming predominantly undergo legitimate reprogramming during fibroblast conversion to iPSCs* refer to this suite of TFs as the reprogrammome. It includes:

1. Transcription factors that were legitimately reprogrammed (their expression levels returned to or exceeded levels observed in the human embryonic stem cells).

2. Transcription factors that were partially reprogrammed. 

3. Transcription factors that were aberrantly reprogrammed.

4. Transcription factors that were irresponsive to reprogramming. 

My initial goal was to determine whether or not specific types of gene ontologies distributed differently amongst the reprogramming categories, in the hope of then developing a better nascent understanding of what pathways the Yamanaka factors are enacting to epigenetically remodel cells. Unfortunately, however, due to some limitations in my data that I will go into further depth on later, I had limited ability to compare the TF expression patterning of my control and OSKM-treated samples to ground-truth human embryonic stem cell data. With my two conditions, then, of RNA sequenced from the same human fibroblast cell line untreated and treated with the OSKM factors, I had to ask different questions. 
I couldn't define specific transcriptomic thresholds after which I could consider any TF expression counts to be officially iPSC-like (successfully reprogrammed), so I tried to boostrap a method of quantifying how much more enriched the set of TFs enriched in iPSCs as compared to fibroblasts (the upreprogramome) were in my data's significantly upregulated TF counts, and how much more enriched the set of TFs enriched in fibroblasts as compared to iPSCs (the downreprogramome) were in my data's significantly downregulated TF counts. My hypothesis was that we would see differential gene expression and GO enrichment patterns between the two conditions. But given that the process of reprogramming fibroblasts to iPSCs is known to take 2-3 weeks, I assumed that - if the upreprogramome and downreprogramome were enriched in upregulated and downregulated genes, respectively - that the degree of enrichment would be mild.  

# Results:

## Dataset limitations
To preface, I will first lay out the limitations of my dataset, both potential and definite. Firstly - although I did my very best to address this by testing for custom BGI adapter sequences when using fastQC, I know that the BGI sequencing platform can sometimes produce peculiar data. Secondly - I have only three replicates per condition. Thirdly, the human ESC samples from the original paper that I intended to use as a reference point for the iPSC-like-ness of my OSKM-treated sample had even fewer replicates, and I was thus not comfortable using it. And finally, differential gene expression among solely transcription factors, and trying to make strong conclusions from TF gene counts at particular time points, can be noisy and risky. TF have short half-lives, can be expressed in short bursts, and are mostly not ubiquitously expressed. I detail my response to this in the methods section.

## Processing the featureCounts count table.

```{r, echo = FALSE, results='hide'}
library(DESeq2)
library(ggplot2)
library(magrittr)
count_table = read.table("finalFeatureCountsOutput.counts")
#print(names(count_table))
print(count_table[1, ])
```


```{r, echo = FALSE, results='hide'}
# Extract the first row as the new column names
new_col_names <- as.character(unlist(count_table[1, ]))

# Modify this line to also capture BJOSKM96HS1 to BJOSKM96HS3
new_col_names[7:12] <- gsub(".*(BJ48hS[1-3]|BJOSKM96hS[1-3]).*\\.bam", "\\1", new_col_names[7:12])

# Assign the new column names to the data frame
names(count_table) <- new_col_names

# Remove the first row from the data frame to eliminate the header row now duplicated in the column names
count_table <- count_table[-1, ]


```

```{r, echo = FALSE, results='hide'}
row.names(count_table) = make.names(count_table$Geneid)

cts_gene_sample <- as.matrix(count_table[, -c(1:6)])
```

```{r, echo = FALSE, results='hide'}
df_coldata <- data.frame(condition = colnames(cts_gene_sample),
row.names = colnames(cts_gene_sample))

```

```{r, echo = FALSE, results='hide'}
# Convert count data in 'cts_gene_sample' from character to numeric
cts_gene_sample <- apply(cts_gene_sample, 2, as.numeric)

# Check if there were any NAs introduced by conversion (which would indicate non-numeric values)
if(any(is.na(cts_gene_sample))) {
  stop("Conversion introduced NAs, indicating non-numeric values in the count data.")
}

# Assuming 'df_coldata' and the design formula are correctly specified
# Try creating the DESeqDataSet object again
dds_yamanaka <- DESeqDataSetFromMatrix(countData = cts_gene_sample,
                                         colData = df_coldata,
                                       rowData =  count_table[,1:6],
                                         design = ~ condition)
```

```{r, echo = FALSE, results='hide'}
keep_genes <- rowSums(counts(dds_yamanaka)) > 0
dds_yamanaka <- dds_yamanaka[keep_genes, ]
```
```{r, echo = FALSE, results='hide'}

conditions <- factor(c("Control", "Control", "Control", 
                       "96 Hours Post-OSKM", "96 Hours Post-OSKM", "96 Hours Post-OSKM"))

# Update colData with the correct conditions
colData(dds_yamanaka)$condition <- conditions

```


```{r, echo = FALSE, results='hide'}
dds_yamanaka <- estimateSizeFactors(dds_yamanaka) #calculate size factors and add them to the object. 
dds_yamanaka <- estimateDispersions(dds_yamanaka)
dds_yamanaka = nbinomWaldTest(dds_yamanaka)

```

```{r, echo = FALSE, results='hide'}
assay(dds_yamanaka, "log_counts") <- log2(counts(dds_yamanaka, normalized = FALSE) + 1)
#Log-normalized read counts
assay(dds_yamanaka, "log_norm_counts") <- log2(counts(dds_yamanaka, normalized=TRUE) + 1)
```

```{r, echo = FALSE, results='hide'}
# Load necessary libraries
library(DESeq2)
library(tidyverse)

# Read the transcription factor list CSV (update path as necessary)
tf_list <- read.csv("Human Transcription Factor List.csv")

# Filter for confirmed transcription factors
confirmed_tfs <- tf_list %>%
  filter(`Is.TF.` == "Yes") %>%
  pull(`Ensembl.ID`)

```

```{r, echo = FALSE, results='hide'}
# Remove version numbers from the rownames of dds_yamanaka
rownames(dds_yamanaka) <- gsub("\\..*", "", rownames(dds_yamanaka))

# Filter dds_yamanaka for TFs
tf_dds <- dds_yamanaka[rownames(dds_yamanaka) %in% confirmed_tfs, ]

```




```{r, echo = FALSE, results='hide'}
# Check if tf_dds has any transcription factors
if (nrow(tf_dds) == 0) {
    stop("No matching transcription factors found.")
} else {

    # Compute the Variance Stabilizing Transformation (VST)
    vst_tf <- vst(tf_dds, blind = FALSE)
    assay(tf_dds, "vst") <- assay(vst_tf)
}
```


## TF PCA Plotting by Condition
The first result that I will demonstrate is just an exploratory PCA on variance-stabilization-transformed TF data. We can see an extremely clear separating between thw two conditions, especially given that the x-axis represents 87% of the variacne, and thus the x-plane-dominated distance between the conditions bespeaks truly large transcriptomic differences. One control sample may appear to vary significantly from the other two, but most of that difference is y-distance, with PC2 only explaining 9% of the variance in the dataset. 
```{r}
plotPCA(vst(tf_dds))
```

## Differential Gene Expression

```{r, echo = FALSE, results='hide'}
all_results <- results(dds_yamanaka, contrast=c("condition", "96 Hours Post-OSKM", "Control"))

```


```{r, echo = FALSE, results='hide'}
# Remove version numbers from the rownames
rownames(dds_yamanaka) <- gsub("\\..*", "", rownames(dds_yamanaka))

# Subset the results to include only transcription factors
tf_results <- all_results[rownames(all_results) %in% confirmed_tfs,]
```


I originally used Benjamini-Hochberg, and while the Yamanaka factors are renownedly transcriptionally transformative, and I thus expect highly differential expression, My general prior is not quite so high as 50% of TFs showing significantly differential expression, and I am still very wary of the erraticness of TF expression patterns. 

```{r, echo = FALSE, results='hide'}
# Recompute p-adjustment only for the subset of transcription factors
tf_results$padj <- p.adjust(tf_results$pvalue, method = "holm")

```


I will illustrate some overall differential gene expression analysis of the two conditions, before delving into specific GO term enrichment analysis 

```{r, echo = FALSE, results='hide'}
tf_results_sorted <- tf_results %>% `[`(order(.$padj),)

library(org.Hs.eg.db)
all_genes <- keys(org.Hs.eg.db, keytype = "ENSEMBL")
symbols <- mapIds(org.Hs.eg.db, keys = all_genes, column = "SYMBOL", keytype = "ENSEMBL", multiVals = "first")
# Also, retrieve ORF names (if available)
orfs <- mapIds(org.Hs.eg.db, keys = all_genes, column = "REFSEQ", keytype = "ENSEMBL", multiVals = "first")

# Add gene symbols or use ENSEMBL ID as fallback
tf_results_sorted$gene_symbol_or_ORF <- ifelse(!is.na(symbols[rownames(tf_results_sorted)]), symbols[rownames(tf_results_sorted)], orfs[rownames(tf_results_sorted)])
tf_results_sorted$ORF <- orfs[rownames(tf_results_sorted)]  # Add ORFs as a separate column
```


```{r, echo = FALSE, results='hide'}
shrunk_tf_results = lfcShrink(dds_yamanaka, coef=2, type = "apeglm")

# Subset the results to include only transcription factors
shrunk_tf_results <- all_results[rownames(shrunk_tf_results) %in% confirmed_tfs,]

# Recompute p-adjustment only for the subset of transcription factors
shrunk_tf_results$padj <- p.adjust(shrunk_tf_results$pvalue, method = "holm")

# Add gene symbols or use ENSEMBL ID as fallback
shrunk_tf_results$gene_symbol_or_ORF <- ifelse(!is.na(symbols[rownames(shrunk_tf_results)]), symbols[rownames(shrunk_tf_results)], rownames(shrunk_tf_results))

# Sort shrunk results
shrunk_tf_results_sorted <- shrunk_tf_results %>% `[`(order(.$padj),)

shrunk_tf_results_sorted$gene_symbol_or_ORF <- ifelse(!is.na(symbols[rownames(shrunk_tf_results_sorted)]), symbols[rownames(shrunk_tf_results_sorted)], orfs[rownames(shrunk_tf_results_sorted)])
shrunk_tf_results_sorted$ORF <- orfs[rownames(shrunk_tf_results_sorted)]  # Add ORFs as a separate column

```


Our MA plot is showing more signifiant results in the direction of upregulation. And, although the difference is not exceptionally significant, we are seeing the expected pattern manifesting of finding more differentially expressed genes as one moves towards the right hand side of the plot, where we find more strongly expressed genes. 
```{r}
plotMA(shrunk_tf_results_sorted, alpha=0.05,
main="LFC Shrunk MA Plot")
```

```{r}
library(EnhancedVolcano)

# Handling p-values of 0 by adjusting to a small number if necessary
shrunk_tf_results_sorted$padj[shrunk_tf_results_sorted$padj == 0] <- min(shrunk_tf_results_sorted$padj[shrunk_tf_results_sorted$padj > 0], na.rm = TRUE) * 10^-1

# Updated EnhancedVolcano call
vp2 <- EnhancedVolcano(shrunk_tf_results_sorted,
                       lab = shrunk_tf_results_sorted$gene_symbol_or_ORF,
                       x = 'log2FoldChange',
                       y = 'padj',
                       pCutoff = 0.05,
                       FCcutoff = 1.5,  # Log2 fold change cutoff
                       title = "Volcano Plot with logFC shrinkage",
                       subtitle = "Highlighting significant changes",
                       col=c('grey30', 'grey30', 'red', 'blue'),  # Colors: non-significant, significant
                       labSize = 3.0,
                       pointSize = c(1.5),  # Ensure this matches the data size; might need adjustment
                       xlim = c(-5, 10))

# Print the plot
print(vp2)
```
This volcano plot further communicates that we are finding more significance among upregulated genes (OSKM-treatment-enriched). Some standouts in terms of significance are KLF4, IRF7, and NR2F2 on the negative L2FC side. 

## Individual gene expression
I look for the top five most positively (higher expression in the 96 hour sample) and negatively expressed genes (lower expression in the 96 hour sample), and report their summaries.
```{r}
# Selecting the top ten most significant genes
top_ten_genes <- head(shrunk_tf_results_sorted, 10)

# Displaying the gene symbols, log fold changes, and adjusted p-values of the top ten genes
top_ten_summary <- top_ten_genes[, c("gene_symbol_or_ORF", "log2FoldChange", "padj")]
print(top_ten_summary)

```
I will then include some example plots of the control-condition comparison of actual normalized expression counts for the top ten most significant down-regulated and up-regulated TF genes. 

```{r, echo = FALSE, results='hide'}
# Assuming shrunk_tf_results_sorted is already sorted by significance
# Filter for significant results and separate positives and negatives
significant_genes <- shrunk_tf_results_sorted[shrunk_tf_results_sorted$padj < 0.05, ]

# Top ten positive log2 fold changes
top_ten_positive <- head(significant_genes[order(significant_genes$log2FoldChange, decreasing = TRUE),], 10)

# Top ten negative log2 fold changes
top_ten_negative <- head(significant_genes[order(significant_genes$log2FoldChange, decreasing = FALSE),], 10)

# Combine the lists for plotting
top_genes_combined <- rbind(top_ten_positive, top_ten_negative)

```


```{r}
# Setup the plotting parameters for five plots per display
par(mfrow=c(1, 5))  # 1 row, 5 columns

# Iterate over the gene list and plot normalized counts
plot_gene_counts <- function(genes_df) {
  for (i in 1:nrow(genes_df)) {
    gene_id <- rownames(genes_df)[i]
    gene_symbol <- genes_df$gene_symbol_or_ORF[i]
    direction <- ifelse(genes_df$log2FoldChange[i] > 0, "Upregulated", "Downregulated")

    plot_title <- paste(direction, gene_symbol, sep=": ")

    # Plot counts for this gene
    plotCounts(dds_yamanaka, gene=gene_id, intgroup="condition", normalized=TRUE, main=plot_title)

    # Refresh the plotting device after every five plots or at the end of the list
    if (i %% 5 == 0 || i == nrow(genes_df)) {
      # This ensures a new image starts if there are more genes to plot
      if (i != nrow(genes_df)) {

      }
    }
  }
}

# Run the plotting function for the top ten positive log fold changes
cat("Displaying Upregulated Transcription Factors:\n")
plot_gene_counts(top_ten_positive)


par(mfrow=c(1, 5))  # Setup for new series of plots
cat("Displaying Downregulated Transcription Factors:\n")
plot_gene_counts(top_ten_negative)



```
## Gene ontology

For purposes of interpretability, however, I am more interested in gene ontology information and how it separates the conditions. The goal is to get a general impression of - assuming that we are seeing signal and not merely the noise of TF's erratic transcriptomic patterning - what biological processes the Yamanaka factors are promoting/inhibiting. 

I made some GO visualizations for the entire set of positive L2FC, statistically significant, 96-hour-sample-enriched genes, and the entire set of negative L2FC, statistically signficant, untreated-fibroblast-enriched genes. 

I begin with the larger sets. 
```{r, echo = FALSE, results='hide'}
library(dplyr)
library(clusterProfiler)
library(org.Hs.eg.db)
library(enrichplot)
library(ggplot2)

# Convert DESeqResults to a standard data frame first
shrunk_tf_results_df <- as.data.frame(shrunk_tf_results_sorted)

# Extract all significant genes with positive and negative log2FoldChanges
all_positive_genes <- shrunk_tf_results_df %>%
                      filter(padj < 0.05 & log2FoldChange > 0) %>%
                      pull(gene_symbol_or_ORF)

all_negative_genes <- shrunk_tf_results_df %>%
                      filter(padj < 0.05 & log2FoldChange < 0) %>%
                      pull(gene_symbol_or_ORF)

# Convert gene symbols to ENTREZ IDs (if your data uses gene symbols)
positive_entrez <- bitr(all_positive_genes, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")
negative_entrez <- bitr(all_negative_genes, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")

```

```{r, echo = FALSE, results='hide'}
# Perform GO enrichment analysis
ego_pos <- enrichGO(gene          = positive_entrez$ENTREZID,
                    OrgDb         = org.Hs.eg.db,
                    keyType       = "ENTREZID",
                    ont           = "BP",
                    pAdjustMethod = "holm",
                    qvalueCutoff  = 0.05)

ego_neg <- enrichGO(gene          = negative_entrez$ENTREZID,
                    OrgDb         = org.Hs.eg.db,
                    keyType       = "ENTREZID",
                    ont           = "BP",
                    pAdjustMethod = "holm",
                    qvalueCutoff  = 0.05)

```

```{r}
# Bar plot of GO terms
barplot(ego_pos, showCategory=10) + ggtitle("Top GO Terms for All Positive Genes")
barplot(ego_neg, showCategory=10) + ggtitle("Top GO Terms for All Negative Genes")
```

```{r, echo = FALSE, results='hide'}
# Extract ENSEMBL gene IDs or gene symbols
positive_genes <- as.character(top_ten_positive$gene_symbol_or_ORF)
negative_genes <- as.character(top_ten_negative$gene_symbol_or_ORF)

```
```{r}
# Convert gene symbols to ENTREZ IDs if necessary
positive_entrez <- bitr(positive_genes, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")
negative_entrez <- bitr(negative_genes, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")

# Perform GO enrichment analysis
ego_pos_ten <- enrichGO(gene          = positive_entrez$ENTREZID,
                    OrgDb         = org.Hs.eg.db,
                    keyType       = "ENTREZID",
                    ont           = "BP",  # Biological Process
                    pAdjustMethod = "holm",
                    qvalueCutoff  = 0.05)

ego_neg_ten <- enrichGO(gene          = negative_entrez$ENTREZID,
                    OrgDb         = org.Hs.eg.db,
                    keyType       = "ENTREZID",
                    ont           = "BP",
                    pAdjustMethod = "holm",
                    qvalueCutoff  = 0.05)

```

```{r}
# Bar plot of GO terms
barplot(ego_pos_ten, showCategory=10) + ggtitle("Top GO Terms for Top 10 Positive Genes")
barplot(ego_neg_ten, showCategory=10) + ggtitle("Top GO Terms for Top 10 Negative Genes")

```

I then made some Revigo treemaps to simplify the GO compositions of these four categories even further, and level them for comparison as much as possible. 

```{r}
# Assuming 'ego_pos' is your enrichGO result for all positive genes
ego_pos_data <- as.data.frame(ego_pos)
revigo_input_pos <- ego_pos_data[, c("ID", "p.adjust")]

# Write this data to a CSV file for uploading to REVIGO
write.csv(revigo_input_pos, "revigo_input_pos.csv", row.names = FALSE)

# Assuming 'ego_pos_ten' is your enrichGO result for the top ten positive genes
ego_pos_ten_data <- as.data.frame(ego_pos_ten)
revigo_input_pos_ten <- ego_pos_ten_data[, c("ID", "p.adjust")]

# Write this data to a CSV file for uploading to REVIGO
write.csv(revigo_input_pos_ten, "revigo_input_pos_ten.csv", row.names = FALSE)

# Assuming 'ego_neg' is your enrichGO result for all negative genes
ego_neg_data <- as.data.frame(ego_neg)
revigo_input_neg <- ego_neg_data[, c("ID", "p.adjust")]

# Write this data to a CSV file for uploading to REVIGO
write.csv(revigo_input_neg, "revigo_input_neg.csv", row.names = FALSE)

# Assuming 'ego_neg_ten' is your enrichGO result for the top ten negative genes
ego_neg_ten_data <- as.data.frame(ego_neg_ten)
revigo_input_neg_ten <- ego_neg_ten_data[, c("ID", "p.adjust")]

# Write this data to a CSV file for uploading to REVIGO
write.csv(revigo_input_neg_ten, "revigo_input_neg_ten.csv", row.names = FALSE)


```

I screenshoted the revigo reduced ontology results, and include them below. Running the code on my .rmd document will download them as a csv to your local computer, however, and you can run the contents through Revigo's querying box if you so desire. I set the genome reference to humans, from amongst their options. 


**Revigo Treemap of all significant positive LFC genes, enriched in the 96 hour condition:**

![](all_pos_treemap.png)
We're seeing some intuitive GO categories, like stem cell population maintenance, but some unexpected ones, like cell fate commitment. Many GO terms indicate TFs that help "regulate" other processes. We don't know whether this is positive or negative regulation, so Revigo in general may yield slightly more inscrutable results with Transcription Factors, given that their annotations in general are not as clear as, say, a protein-encoding gene. 

**Revigo Treemap of all significant negative LFC genes, enriched in the untreated fibroblast condition:**

![](all_neg_treemap.png)
The embryonic organ development" semantic group, which very much dominates the GO composition, is aggregating developmental/cell differentiation processes, for the most part. This is very much as to be expected if the trajectory of Yamanaka factor induction is meant to take a fibroblast from differentiated to stem-like. 


**Revigo Treemap of top ten most significant positive LFC OSKM-treatment-enriched genes:**

![](pos_ten_treemap.png)

**Revigo Treemap of top ten most significant negative LFC untreated-fibroblast-enriched genes:**

Gastrulation is a crucial phase early in the embryonic development of animals, during which the single-layered blastula is reorganized into a multilayered structure known as the gastrula. It's essentially the process by which the embryo transforms from a simple spherical ball of cells into a structure with multiple layers and a defined axis, which will give rise to distinct tissues and organs. Figuring this out, and seeing this so enriched in the most significantly upregulated genes in the whole TF dataset, confused me at first. This is a process by which embryos transform out of their initial shape, which made me immediately assume that it was more in the category of cell fate determination and differentiation. However, given that the Yamanaka factors revert differentiated cells back to iPSCs in a continuous manner, it makes sense that we'd see an enrichment for embryonic traits - even if the phenotypes they describe transition embryos away from their embryonic-ness. 

![](neg_ten_treemap.png)
This group is small enough that Revigo is not spectacularly enlightening, but the single largest GO term being downregulated is cell fate commitment, further in-keeping with our intuitions. Interestingly, however, "cell fate commitment" ontologies are listed very prominently amongst the top ten most significantly untreated-fibroblast-enriched genes. Yet it's also a sizable ontology among the greater set of 96-hour-enriched genes. I can imagine, however, how cells becoming more stem-like could involve downregulating TFs that help preserve a current cell fate trajectory *and* upregulating TFs that have to be more active in steering a more embryonic cell towards an ultimate cell fate (and are thus markers of a more stem-like cell).

And finally, I proceed to the GSEA portion of my results.

I now load a table from the supplementary information of the paper I have most referenced, that includes information that will be the basis for my constructing the gene sets of the "upreprogramome" and "downreprogramome" (see introduction). They compared human fibroblasts and human embryonic stem cells TF counts to establish each's relative patterning. And while I wasn't comfortable using their hESC data in direct comparison with mine (to establish whether or not any of my TF gene counts appear iPSC-like) given that it was generated with subtle differences in QC and processing from mine, I was comfortable extracting the list of TFs that they assigned to the upreprogamome and downreprogamome to use as gene sets in GSEA for my own case-control study. 

And again, for clarity's sake:

1. The TF upreprogramome refers to transcription factors that have to be upregulated in the fibroblast for pluripotency induction.

2. The TF downreprogramome refers to fibroblast-enriched and specific transcriptional factors that have to be downregulated for pluripotency. 


```{r, echo = FALSE, results='hide'}
library(readxl)
data <- read_excel("paper_log_fold_table.xlsx")
print(paste0("The length of my metadata table is ",nrow(data)))
```
## Processing the upreprogramome and downreprogramome in

```{r, echo = FALSE, results='hide'}
# Assuming the columns are named exactly as "ensembl", "log2FoldChange", "padj", "Fibroblast_mean", and "ESC_mean" in the Excel file.
selected_data <- data %>%
  select(ensembl, log2FoldChange, padj, Fibroblast_mean, ESC_mean)

```


```{r, echo = FALSE, results='hide'}
# Assuming selected_data is already loaded and is a dataframe

# Load the necessary library
library(dplyr)

# Ensure that log2FoldChange and padj are numeric
selected_data <- selected_data %>%
  mutate(
    log2FoldChange = as.numeric(as.character(log2FoldChange)),
    padj = as.numeric(as.character(padj))
  ) %>%
  # Handle NAs that may have been introduced by coercion
  filter(!is.na(log2FoldChange) & !is.na(padj))

# Define the thresholds used in the original paper - Human Transcription Factor Responsive to Initial Reprogramming Predominantly Undergo Legitimate Reprogramming during Fibroblast Conversion to iPSCs. 
active_threshold <- 50
fold_change_threshold <- 2
q_value_threshold <- 0.01

upreprogramome <- selected_data %>%
  filter(ESC_mean > active_threshold, log2FoldChange > fold_change_threshold,
         padj < q_value_threshold)

downreprogramome <- selected_data %>%
  filter(Fibroblast_mean > active_threshold, log2FoldChange < -fold_change_threshold,
         padj < q_value_threshold)


cat("Upreprogramome:", nrow(upreprogramome), "\n")
cat("Downreprogramome:", nrow(downreprogramome), "\n")


```


```{r, echo = FALSE, results='hide'}
library(fgsea)


upreprogramome_genes <- upreprogramome$ensembl
downreprogramome_genes <- downreprogramome$ensembl


# Create a list of gene sets for GSEA
gene_sets <- list(
  Upreprogramome = upreprogramome_genes,
  Downreprogramome = downreprogramome_genes
)

```


```{r, echo = FALSE, results='hide'}
# Extract all significant genes with positive and negative log2FoldChanges
significant_up_genes <- shrunk_tf_results_df %>%
                      filter(padj < 0.05 & log2FoldChange > 0)

significant_down_genes <- shrunk_tf_results_df %>%
                         filter(padj < 0.05 & log2FoldChange < 0)


```


```{r, echo = FALSE, results='hide'}
# Assuming shrunk_tf_results_df is your DESeq2 results dataframe and upreprogramome_genes is a vector of gene IDs
# First, ensure that your gene IDs in upreprogramome_genes do not have version numbers if they are present in your shrunk_tf_results_df
upreprogramome_genes <- sub("\\..*", "", upreprogramome_genes)

# Now, subset your shrunk_tf_results_df to include only the rows with gene IDs that match the upreprogramome_genes
upreprogramome_genes <- shrunk_tf_results_df[rownames(shrunk_tf_results_df) %in% upreprogramome_genes, ]

# Assuming shrunk_tf_results_df is your DESeq2 results dataframe and upreprogramome_genes is a vector of gene IDs
# First, ensure that your gene IDs in upreprogramome_genes do not have version numbers if they are present in your shrunk_tf_results_df
downreprogramome_genes <- sub("\\..*", "", downreprogramome_genes)

# Now, subset your shrunk_tf_results_df to include only the rows with gene IDs that match the upreprogramome_genes
downreprogramome_genes <- shrunk_tf_results_df[rownames(shrunk_tf_results_df) %in% downreprogramome_genes, ]

```

```{r, echo = FALSE, results='hide'}

# Extract gene IDs as vectors from the filtered dataframes
significant_up_gene_ensembl <- rownames(significant_up_genes)
significant_down_gene_ensembl <- rownames(significant_down_genes)

upreprogramome_gene_ensembl <- rownames(upreprogramome_genes)
downreprogramome_gene_ensembl <- rownames(downreprogramome_genes)


```


```{r, echo = FALSE, results='hide'}
# Assuming that upreprogramome_genes and downreprogramome_genes are correctly formatted as vectors
# Now the lists for the Venn diagram are set up correctly:
significant_up_gene_ensembl <- as.character(significant_up_gene_ensembl)
significant_down_gene_ensembl <- as.character(significant_down_gene_ensembl)

upreprogramome_gene_ensembl <- as.character(upreprogramome_gene_ensembl)
downreprogramome_gene_ensembl <- as.character(downreprogramome_gene_ensembl)
```


```{r}
library(VennDiagram)
library(grid)

# Define dimensions in inches
width_in_inches <- 10
height_in_inches <- 6

# Create Venn Diagrams and save as PNG files
venn.plot.up <- venn.diagram(
  x = list(
    Significant_Up = significant_up_gene_ensembl,
    Upreprogramome = upreprogramome_gene_ensembl
  ),
  category.names = c("Significantly Upregulated", "Upreprogramome"),
  filename = "venn_diagram_up.png",  # Saving as PNG file
  output = TRUE,
  imagetype = "png",
  height = height_in_inches,  # height in inches
  width = width_in_inches,    # width in inches
  resolution = 300,
  compression = "lzw",
  units = "in",  # specify units
  lwd = 2,       # line width
  cex = 1.5,     # font size scaling factor
  fontface = "bold",  # font style
  cat.cex = 1.2,      # category label scaling factor
  cat.fontface = "bold",  # category label font style
  cat.default.pos = "outer",  # category label position
  cat.dist = 0.05,   # distance for category labels from the center of the diagram
  margin = 0.05      # margin around the plot
)

venn.plot.down <- venn.diagram(
  x = list(
    Significant_Down = significant_down_gene_ensembl,
    Downreprogramome = downreprogramome_gene_ensembl
  ),
  category.names = c("Significant Downregulated", "Downreprogramome"),
  filename = "venn_diagram_down.png",  # Saving as PNG file
  output = TRUE,
  imagetype = "png",
  height = height_in_inches,
  width = width_in_inches,
  resolution = 300,
  compression = "lzw",
  units = "in",
  lwd = 2,
  cex = 1.5,
  fontface = "bold",
  cat.cex = 1.2,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.dist = 0.05,
  margin = 0.05
)

```
![](venn_diagram_down.png)


![](venn_diagram_up.png)

I generate simple venn diagrams for a clear first introduction to the overlap of the up/down-reprogamome and our significantly upregulated/downregulated control/OSKM data. The down venn diagram has 58 overlapping TFs, and the up venn diagram has 46. Obviously, this visualization is agnostic to a TF's degree of L2FC and p value. 

## GSEA
```{r, echo = FALSE, results='hide'}
# Load the required library
library(fgsea)

# Prepare your ranked list of genes
# This example assumes `shrunk_tf_results_df` is your dataframe from DESeq2 results
# and it's sorted from most upregulated to most downregulated based on log2FoldChange
ranked_list <- with(shrunk_tf_results_df, setNames(-log10(padj) * sign(log2FoldChange), shrunk_tf_results_df$gene_symbol_or_ORF))


```


```{r}
# Prepare gene sets from the filtered dataframes
gene_sets <- list(
  Upreprogramome = upreprogramome_genes$gene_symbol_or_ORF,
  Downreprogramome = downreprogramome_genes$gene_symbol_or_ORF
)

# Run GSEA for upregulated genes
# 'scoreType = "pos"' focuses the analysis on the positive side of the ranked list
gsea_results_up <- fgsea(pathways = gene_sets["Upreprogramome"], 
                         stats = ranked_list, 
                         scoreType = "pos",
                         minSize = 15,    # Minimum size of gene set to consider
                         maxSize = 500,   # Maximum size of gene set to consider
                         nperm = 10000)   # Number of permutations for estimation

# Run GSEA for downregulated genes
# 'scoreType = "neg"' focuses the analysis on the negative side of the ranked list
gsea_results_down <- fgsea(pathways = gene_sets["Downreprogramome"], 
                           stats = ranked_list, 
                           scoreType = "neg",
                           minSize = 15,
                           maxSize = 500,
                           nperm = 10000)

# Print the GSEA results to examine the outputs
print(gsea_results_up)
print(gsea_results_down)
```


```{r}
# Visualization of the most significant GSEA results
# Ensure the pathway is selected correctly

# Since 'most_significant_up$pathway' and 'most_significant_down$pathway' return the names directly:
plotEnrichment(pathway = gene_sets[["Upreprogramome"]], ranked_list)
# Add a title manually using the title() function in base R


plotEnrichment(pathway = gene_sets[["Downreprogramome"]], ranked_list)
# Add a title manually using the title() function in base R

```
These plots extend over my transcription factors ranked from left to right by most upregulated to most downregulated. 

The first of these two enrichment plots is for the upreprogramome across my TFs, and the second is for the downreprogramome.


From viewing the summary of the GSEA objects, the leadingEdge gene that contributes most to the enrichment score of the upreprogramome before it peaks is POU5F1, which drove the gastrulation GO term in the original DGE analysis. It plays a key and well characterized role in embryonic development and stem cell pluripotency. The leadingEdge gene that contributed most to the enrichment score of the downreprogramome before it peaked was NR2F2. NR2F2 has widely-varying purposes, but importantly has been implicated in cell differentiation regulation. Its downregulation driving a downregulome enrichment stands to reason, then.

And overall, these significance values for the enrichment of the upreprogramome in upregulated genes and the downreprogramome in downregulated genes are very low, well below, the significance cutoff, and even lower for the upreprogramome.

# Methods:

## Conditions
The RNA sequencing data used in this analysis was derived from human fiborblast cells (the BJ line) at 48 hours post-seeding with no OSKM treatment (control), and the same type of fibroblast 96 hours after seeding and OSKM transduction. I found the samples from the Kejin Hu Nature paper that I have referenced so frequently. 

## Whole Process Overview
Fastq files were downloaded using prefetch from the SRA database and processed with fasterq-dump. Quality control was performed using FastQC and MultiQC, followed by alignment using STAR with default parameters and subsequent quality checks with SAMtools and QoRTs. The library was stranded and the reads were paired-end. The data being BGI, I provided FastQC with custom BGI adapter sequences I found online in .txt file, which was necessary given that FastQC by default seeks out Illumina sequencers. I chose QoRTs after repeatedly sustaining intractable error message with RSeQC. A high duplicate rate was initially concerning, but ~26M aligned read pairs per sample greatly ameliorated this issue. The aligned reads were quantified using featureCounts, and differential expression analysis was conducted with DESeq2. I used clusterProfiler and Revigo for many of my GO visualization, and the fgsea library for the gene set enrichment analysis that I concluded with. Throughout the analysis, reproducibility and data integrity were prioritized, as all bash scripts were version-controlled and shared on GitHub.







# Discussion:

## Couldn't evaluate reprogramming legitimacy
As aforementioned, my greatest and most higher level difficulty was not being able to evaluate reprogramming legitimacy for TFs on account of the hESC cell line from the same experiment having two few technical replicates. And again, my solution was load a table from the reprogramming legitimacy paper that includes information that served as the basis for my constructing the gene sets of the "upreprogramome" and "downreprogramome" (see introduction). They compared human fibroblasts and human embryonic stem cells TF counts to establish each ones' relative patterning. And while I wasn't comfortable using their hESC data in direct comparison with mine (to establish whether or not any of my TF gene counts appear iPSC-like and had been "legitimately reprogrammed") given that it was generated with subtle differences in QC and processing from mine, I was comfortable extracting the list of TFs that they assigned to the upreprogamome and downreprogamome to use as gene sets in GSEA for my control versus OSKM comparison. 

## Normalization questions for TF subsetting
As far as real technical quandaries are concerned, I wasn't certain initially if I should recompute size factors for my gene set after subsetting for just transcription factors, or keep the TF counts normalization relative to the whole transcriptome. Following Merv's advice, I wanted to evaluate how dramatically the size factors would change if I recomputed post-subsetting, as well as whether or not any outlier high expression value TFs among a fairly small total TF count sample size could mess up the normalization.

I proceeded testing this below: 

This plot, besides the ultimate hubris of not subsetting directly from a results object of dds_yamanaka, is useful. Knowing how much TFs contribute to overall expression can give us a sense of whether or not focusing exclusively on them would require specialized processing, and hwo much expression stability can be expected from them.
```{r}
# Calculate total counts per sample in the original dataset
total_counts_all_genes <- colSums(counts(dds_yamanaka))

# Calculate total counts per sample for just transcription factors
total_counts_tfs <- colSums(counts(tf_dds))

# Calculate the fraction of counts from TFs
fraction_tfs <- total_counts_tfs / total_counts_all_genes

# Print the fraction of counts from TFs
print(fraction_tfs)

# Create a barplot of the fraction of counts from TFs
barplot(fraction_tfs, 
        main = "Fraction of Counts from Transcription Factors", 
        ylab = "Fraction", 
        xlab = "Samples", 
        col = "blue", 
        las = 2)  # Adjust label orientation if necessary

```
The fraction of all counts per sample that are coming from human transcriptions factors ranges from 4.99% to 6.09%. This is, self-evidently, quite a small fraction of the whole gene set. 


Following Merv's advice from Piazza, I want to evaluate how my size factors change from the original ones for the entire gene set, versus recomputed ones for just my 1,549 human transcription factors.

```{r, echo = FALSE, results='hide'}
sizeFactors(dds_yamanaka)
```

```{r, echo = FALSE, results='hide'}
sizeFactors(tf_dds)
```

```{r}
# Create a copy of the original DESeq2 object
tf_dds_copy <- tf_dds

# Recalculate size factors for just the transcription factors
tf_dds_copy <- estimateSizeFactors(tf_dds_copy)

sizeFactors(tf_dds_copy)
```

```{r}
# Extract size factors
original_size_factors <- sizeFactors(tf_dds)
new_size_factors <- sizeFactors(tf_dds_copy)

# Create a plot for comparison
plot(original_size_factors, new_size_factors, 
     xlab = "Original Size Factors",
     ylab = "New Size Factors for TFs",
     main = "Comparison of Size Factors",
     pch = 19, col = "blue")
abline(0, 1, col = "red")  # Add a y=x line to aid in visual comparison

```
 The $x = y$ line here represents where points would fall if we observed no size factor change. The first three (reading from left to right) dots represent my pre-Yamanaka samples, and the last three represent my post-Yamanaka samples. The pre-yamanaka size factors decrease, as compared to the original, and the post-ones increase. 

My concern is that a small number of TFs, expressed highly enough, could make everything else look down-regulated unfairly. So I use Cook's distance, which measures the influence of each TF gene on the computed size factors, to see if I have any serious outliers shifting the balance of my normalization.


```{r}

# Calculate Cook's distances for each gene
dds_fit <- DESeq(tf_dds_copy, fitType="mean")
cooks_distances <- assays(dds_fit)[["cooks"]]
mean_cooks <- rowMeans(cooks_distances, na.rm = TRUE)

# Plot Cook's distances to identify highly influential genes
plot(mean_cooks, type='h', 
     main="Cook's Distance for Transcription Factors",
     xlab="Transcription Factor Index", ylab="Cook's Distance",
     col="red")

```
Purely visually, we can see that two to three different transcription factor genes have vastly higher Cook's Distance scores than the rest.


```{r}
# Determine the 99th percentile threshold for Cook's distance
threshold_cooks <- quantile(mean_cooks, 0.99)

# Identify TFs whose Cook's distance exceeds the 95th percentile
high_influence_tfs <- names(mean_cooks)[mean_cooks > threshold_cooks]

# Report Cook's distances for these TFs
high_influence_cooks <- mean_cooks[high_influence_tfs]

# Create a named vector to display results more clearly
results <- setNames(high_influence_cooks, high_influence_tfs)

# Print the results
print(results)
```
Setting a 99th percentile cutoff, we see that, even among the top 1th percentile of influential TF genes, we are seeing one TF (ENSG00000168874/ATOH8), that contributes almost 4 times more to size factor calculation than the bottom of this quantile.

I am deciding that this is enough evidence for not redoing normalization with size factors specific to just the TF genes. I will adjust my p-values in my transcription factor DESeq2 object to just reflect the sample size of the TF gene counts, however. 

## Multiple testing method choice:
Knowing that - among TFs - the differential expression normalization assumption that most genes are not differentially expressed between conditions isn't always true with TFs, I also wanted to choose a stringent p value adjustment method. BH - which, from the literature, seems to be something of a default with RNA-seq data, was finding approximately 50% of all of my TFs to be significantly differentially expressed. Given OSKM iPSC induction takes two to three weeks, that was much higher than my prior expectation, so I ended up going with the Holm-Bonferroni correction from some online research. 

Even with the relative stringency of the Holm-Bonferroni correction, we are still observing 17.7% of our TFs as being significantly differentially expressed. 

## Choosing BP for my GO anlysis:
From some online research, BP was volunteered as best covering most of the functional categories I would be looking for among TFs after OSKM induction. 


# Data Summary Table:

| Dataset Description                          | Source | Key Characteristics and Findings |
|----------------------------------------------|--------|----------------------------------|
| **Raw sequencing data**                     | SRA Database | Human fibroblast cells (BJ line), control and OSKM-treated samples at different time points post-seeding. |
| **Quality control checks**                  | FastQC, MultiQC | Custom BGI adapter sequences used due to BGI sequencing platform; high duplicate rate noted but mitigated by high aligned read pairs. |
| **Alignment and quantification of reads**   | STAR, featureCounts | Paired-end reads, stranded library; reads aligned to human hg38 reference genome, quantified using featureCounts. |
| **Differential expression analysis**        | DESeq2 | Analysis conducted with three replicates per condition; adjustment for multiple comparisons using Holm-Bonferroni method. |
| **Gene Ontology (GO) enrichment analysis**  | CP, Revigo | Used to understand biological processes promoted or inhibited by Yamanaka factors. |
| **Gene Set Enrichment Analysis (GSEA)**     | fgsea | Identification of significantly upregulated (upreprogramome) and downregulated (downreprogramome) gene sets in relation to Yamanaka factor induction. |
| **Venn diagrams**                           | VD | Visualization of overlap between up/down-reprogramome and significantly regulated genes in control/OSKM data. |
| **Principal Component Analysis (PCA)**      | DESeq2 `plotPCA` | Clear separation between conditions observed; significant variation attributed to treatment. |
| **Enhanced Volcano plots**                           | Volcano | Visualization of differential gene expression, highlighting significant changes post-OSKM treatment. |
| **Cook's distance assessment**              | DESeq2 `cooks` | Influential gene detection to evaluate impact on normalization; one gene (ATOH8) notably influential. |



# Bibliography:
Cevallos, Ricardo R., et al. “Human Transcription Factors Responsive to Initial Reprogramming Predominantly Undergo Legitimate Reprogramming during Fibroblast Conversion to Ipscs.” Nature News, Nature Publishing Group, 12 Nov. 2020, www.nature.com/articles/s41598-020-76705-y. 

Hochedlinger, Konrad. “A Molecular Roadmap of Reprogramming Somatic Cells into iPS Cells.” Cell, www.cell.com/cell/pdf/S0092-8674(20)30610-3.pdf?_returnURL=https:/linkinghub.elsevier.com/retrieve/pii/S0092867420306103?showall=true. Accessed 23 Apr. 2024. 

Lambert, Samuel A. Cell, Cell, www.cell.com/cell/pdf/S0092-8674(20)30610-3.pdf?_returnURL=https:/linkinghub.elsevier.com/retrieve/pii/S0092867420306103?showall=true. Accessed 23 Apr. 2024. 

Mikkelsen, Tarjei S., et al. “Dissecting Direct Reprogramming through Integrative Genomic Analysis.” Nature News, Nature Publishing Group, 28 May 2008, www.nature.com/articles/nature07056. 


